### ISSUE TITLE
cuDNN RNN doesn't support multiple backwards

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#703]
It always deletes `self._nested_output` regardless of `retain_variables`

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-14-8a13d1ccf570> in <module>()
----> 1 loss.backward(tmp)

/data/mulga/trask/anaconda3/lib/python3.6/site-packages/torch/autograd/variable.py in backward(self, gradient, retain_variables)
    156                     'or with gradient w.r.t. the variable')
    157             gradient = self.data.new().resize_as_(self.data).fill_(1)
--> 158         self._execution_engine.run_backward((self,), (gradient,), retain_variables)
    159 
    160     def register_hook(self, hook):

/data/mulga/trask/anaconda3/lib/python3.6/site-packages/torch/autograd/function.py in backward(self, *gradients)
    206 
    207     def backward(self, *gradients):
--> 208         nested_gradients = _unflatten(gradients, self._nested_output)
    209         del self._nested_output
    210         result = self.backward_extended(*nested_gradients)
```

**Comment by apaszke** [comment]
@ngimel we'd have to clone the scratch space from before the backward if we want to do it second time, right?

**Comment by ngimel** [comment]
You never need to preserve scratch space (it's called workspace in cudnn API). You need reserveSpace that was computed in forward for backward, in rnn.py it is called fn.reserve
https://github.com/pytorch/pytorch/blob/master/torch/backends/cudnn/rnn.py#L281
You don't need to clone it as it is not modified in backward.

**Comment by apaszke** [comment]
Can we rely on that? Or might it change in the future versions? The cuDNN user guide says:

> The data in reserveSpace must have previously been generated by cudnnRNNForwardTraining. The same reserveSpace data must be used for future calls to cudnnRNNBackwardWeights if they execute on the same input data.

Also (description of `cuDNNBackwardWeights`):

> The data in reserveSpace must have previously been generated by cudnnRNNBackwardData.

That sounds to me as if you always have to call `cudnnRNNBackwardData` before `cudnnRNNBackwardWeights`, because it modifies something. If that's not true, can we skip `cudnnRNNBackwardData` altogether if the input doesn't require grad?

**Comment by ngimel** [comment]
Ah, you are right, my mistake (should not have relied on memory). In backwardData it is 
```                                              
void * reserveSpace,
``` 
so it is read/write, and only in backwardFilter it is 
```
const void * reserveSpace,
```

**Comment by apaszke** [comment]
So, just to be sure - we need to clone the `reserveSpace` if we know that we want to backprop multiple times, correct?

**Comment by ngimel** [comment]
Yes.

**Comment by apaszke** [comment]
Fixed in #734.
